### ISSUE TITLE
Dyno has long time out 36s

### ISSUE THREAD

**Issue by diegopacheco** [Netflix/dyno#149]
@ipapapa @timiblossom 

Dyno has long timeout default looks to be 36s. If I try to connect on the wrong IP where Dynomite is not there it takes 36s to dyno realize it's a bad IP.

``` bash
$ ./gradlew run
:compileJava
:processResources UP-TO-DATE
:classes
:run
log4j:WARN No appenders could be found for logger (com.netflix.config.sources.URLConfigurationSource).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
com.netflix.dyno.connectionpool.exception.PoolOfflineException: PoolOfflineException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: null, datacenter: null, status: Down], latency=0(0), attempts=0]host pool is offline and no Racks available for fallback
--
--
TIME TO RUN: 36 seconds
> Building 75% > :run
```

Dyno does not respect timeout parameters such as: setConnectTimeout, setFailOnStartupIfNoHostsSeconds, setMaxTimeoutWhenExhaustedm setSocketTimeout... Sample code: 

``` bash
        DynoJedisClient dynoClient = new DynoJedisClient.Builder()
                .withApplicationName("MY_APP")
                    .withDynomiteClusterName("MY_CLUSTER")
                    .withCPConfig( new ArchaiusConnectionPoolConfiguration("MY_APP")
                                        .setPort(8101)
                                        .withTokenSupplier(testTokenMapSupplier)
                                        .setMaxConnsPerHost(100) 

                                        .setConnectTimeout(5)
                                        .setFailOnStartupIfNoHosts(true)
                                        .setFailOnStartupIfNoHostsSeconds(5)
                                        .setMaxTimeoutWhenExhausted(5)
                                        .setSocketTimeout(5)
                                        .setRetryPolicyFactory(new RetryNTimes.RetryFactory(2))
                    )
                    .withHostSupplier(customHostSupplier)
                    .build();
```

The whole code is here: https://github.com/diegopacheco/netflixoss-pocs/tree/master/dyno-timout

I think would be nice is Dyno respect that timeouts and we could configure it. 36s is a very long timeout.

Cheers,
Diego Pacheco

**Comment by jcacciatore** [comment]
It does respect the timeouts, you are not setting them properly. I will send a code snippet this afternoon

**Comment by diegopacheco** [comment]
Thanks @jcacciatore 

I'using dyno 1.4.7. I try to update to dyno 1.5.1 and realise there is a new property: setPoolShutdownDelay which a set to ZERO but the problem persists to me 36s timeout.

Looking forward to your snippet.

**Comment by jcacciatore** [comment]
All you should need to do is replace ```ArchaiusConnectionPoolConfiguration``` with ```ConnectionPoolConfigurationImpl``` otherwise your overrides will not be applied. I believe the 36 seconds you are observing comes from the default socket connect timeout of 12 seconds which gets retried twice.

Another thing to note is that the timeouts are all specified in millis except where 'Seconds' is included in the name (e.g. ```setFailOnStartupIfNoHostsSeconds```). I mention this b/c 5 millisecond connect timeout will surely fail, unless you are setting it this low simply for the purposes of your testing.

The snippet below is a copy+paste of yours using the correct config class and sane timeouts. Let me know if that doesn't work for you.

```java
DynoJedisClient dynoClient = new DynoJedisClient.Builder()
                .withApplicationName("MY_APP")
                    .withDynomiteClusterName("MY_CLUSTER")
                    .withCPConfig( new ConnectionPoolConfigurationImpl("MY_APP")
                                        .setPort(8101)
                                        .withTokenSupplier(testTokenMapSupplier)
                                        .setMaxConnsPerHost(100) 
                                        .setConnectTimeout(5000)
                                        .setFailOnStartupIfNoHosts(true)
                                        .setFailOnStartupIfNoHostsSeconds(5)
                                        .setMaxTimeoutWhenExhausted(5)
                                        .setSocketTimeout(5000)
                                        .setRetryPolicyFactory(new RetryNTimes.RetryFactory(2))
                    )
                    .withHostSupplier(customHostSupplier)
                    .build();
```

**Comment by diegopacheco** [comment]
@jcacciatore 

Thank you now it works!

I changed from ArchaiusConnectionPoolConfiguration to ConnectionPoolConfigurationImpl.

```bash
$ ./gradlew run
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:run
log4j:WARN No appenders could be found for logger (com.netflix.dyno.jedis.DynoJedisClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
com.netflix.dyno.connectionpool.exception.NoAvailableHostsException: NoAvailableHostsException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: null, datacenter: null, status: Down], latency=0(0), attempts=0]Could not find any remote Racks for fallback
--
--
TIME TO RUN: 3 seconds
> Building 75% > :run
```

However ArchaiusConnectionPoolConfiguration does not work. So we should never use this class? Are you guys planning to remove this class or fix the timeouts there? 

Cheers,
Diego Pacheco.

**Comment by jcacciatore** [comment]
Glad it is working for you @diegopacheco. I admit, the way the configuration code is currently wired up is not intuitive and could use some work.

ArchaiusConnectionPoolConfiguration does work, however, and it's purpose is to hook into the Netflix configuration management APIs. This allows internal users here to update properties via centralized internal tools rather than programmatically updating their code for a property change. This is desirable since building, baking an AMI, and deploying code is a long extensive process for updating a property.

Once the code is changed to utilize Archaius v2 it may be more appealing for external users since you can configure a poller that periodically scans your property source(s) and Archaius handles updating changed property values in a thread safe way while your application is running. This is desirable, for example, if you want to change your max connections from 5 to 25 which currently requires an application restart.

Hope that Helps,
Jason

**Comment by jcacciatore** [comment]
Closing this issue since it is working for you. If there is something additional feel free to re-open it or start a new issue.

**Comment by diegopacheco** [comment]
@jcacciatore 

I think we might have an issue.

When we use ArchaiusConnectionPoolConfiguration the timeout is 36 seconds.

When we use ConnectionPoolConfigurationImpl the timeout is 2 seconds.

No matter what I pass on the config is always this times.

Sample:
```bash
		DynoJedisClient dynoClient = new DynoJedisClient.Builder()
					.withApplicationName("MY_APP")
		            .withDynomiteClusterName("MY_CLUSTER")
		            .withCPConfig( new ConnectionPoolConfigurationImpl("MY_APP")
		            					.setPort(8101)
		            					.withTokenSupplier(testTokenMapSupplier)
		            					.setMaxConnsPerHost(100) 
		            					
		            					.setPoolShutdownDelay(53000)
		            					.setConnectTimeout(5)
		            					.setFailOnStartupIfNoHosts(true)
		            					.setFailOnStartupIfNoHostsSeconds(55)
	                					.setMaxTimeoutWhenExhausted(55000)
		            					.setSocketTimeout(55000)
		            					.setRetryPolicyFactory(new RetryNTimes.RetryFactory(53))
		            )
		            .withHostSupplier(customHostSupplier)
		            .build();
```

Result:
```bash
log4j:WARN No appenders could be found for logger (com.netflix.dyno.jedis.DynoJedisClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
com.netflix.dyno.connectionpool.exception.NoAvailableHostsException: NoAvailableHostsException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: null, datacenter: null, status: Down], latency=0(0), attempts=0]Could not find any remote Racks for fallback
--
--
TIME TO RUN: 2 seconds
``` 

Cheers,
Diego Pacheco

**Comment by jcacciatore** [comment]
These *should* all be documented in the [ConnectionPoolConfiguration](https://github.com/Netflix/dyno/blob/master/dyno-core/src/main/java/com/netflix/dyno/connectionpool/ConnectionPoolConfiguration.java) interface, and most of them are however I will file an issue and clean this up myself.

Also, it turns out that there is a minor bug I found while double checking the settings are applied correctly, specifically ```setSocketTimeout()``` is not being used. Currently both the connect timeout and read timeout on the socket are controlled by ```setConnectTimeout()```. I will file a separate issue for this and fix it up as well.

It might help if you explain what you are trying to achieve or test with this code ?

**Comment by jcacciatore** [comment]
See [PR 152](https://github.com/Netflix/dyno/pull/152). Let me know if anything isn't clear

**Comment by diegopacheco** [comment]
@jcacciatore 

Thanks for the support. The docs looks good to me. Besides the bug the main problem for me is that the timeouts are not consistent with the configs. 

My expectation is if a SET .setConnectTimeout(2000) That I got the timeout in 2s or 4s if the number of retrys is 1, 6s if the number of retry is 3 and so on and on... 

The code is consistent but is not the consistency I was expecting. For instance consider the following ConnectionTimeout VS the execution time.

1 -> 0s
2 -> 1s
4 -> 2s
8 -> 3s
16 -> 6s

IMHO all this should be 0 since the value should be in milliseconds. I was expecting pass the value in milliseconds like 2000 get the timeout in 2s * number of retrys. 

However this is not happening. Take a look on this samples:

This is not a BIG problem since i can manage to get the timeout out a want but the code is not intuitive and does not feel right to me. 

Here is the proof of what I'm talking about... 

```bash
```

```bash
		DynoJedisClient dynoClient = new DynoJedisClient.Builder()
					.withApplicationName("MY_APP")
		            .withDynomiteClusterName("MY_CLUSTER")
		            .withCPConfig( new ConnectionPoolConfigurationImpl("MY_APP")
		            					.setPort(8101)
		            					.withTokenSupplier(testTokenMapSupplier)
		            					.setMaxConnsPerHost(100) 
		            					
		            					//.setPoolShutdownDelay(0)
		            					.setConnectTimeout(1)
		            					//.setFailOnStartupIfNoHosts(true)
		            					//.setFailOnStartupIfNoHostsSeconds(12)
	                					//.setMaxTimeoutWhenExhausted(12)
		            					//.setSocketTimeout(12)
		            					.setRetryPolicyFactory(new RetryNTimes.RetryFactory(2))
		            )
		            .withHostSupplier(customHostSupplier)
		            .build();
```
```bash
log4j:WARN No appenders could be found for logger (com.netflix.dyno.jedis.DynoJedisClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
com.netflix.dyno.connectionpool.exception.NoAvailableHostsException: NoAvailableHostsException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: null, datacenter: null, status: Down], latency=0(0), attempts=0]Could not find any remote Racks for fallback
--
--
TIME TO RUN: 0 seconds
```

```bash
		DynoJedisClient dynoClient = new DynoJedisClient.Builder()
					.withApplicationName("MY_APP")
		            .withDynomiteClusterName("MY_CLUSTER")
		            .withCPConfig( new ConnectionPoolConfigurationImpl("MY_APP")
		            					.setPort(8101)
		            					.withTokenSupplier(testTokenMapSupplier)
		            					.setMaxConnsPerHost(100) 
		            					
		            					//.setPoolShutdownDelay(0)
		            					.setConnectTimeout(2)
		            					//.setFailOnStartupIfNoHosts(true)
		            					//.setFailOnStartupIfNoHostsSeconds(12)
	                					//.setMaxTimeoutWhenExhausted(12)
		            					//.setSocketTimeout(12)
		            					.setRetryPolicyFactory(new RetryNTimes.RetryFactory(2))
		            )
		            .withHostSupplier(customHostSupplier)
		            .build();
```
```bash
log4j:WARN No appenders could be found for logger (com.netflix.dyno.jedis.DynoJedisClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
com.netflix.dyno.connectionpool.exception.NoAvailableHostsException: NoAvailableHostsException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: null, datacenter: null, status: Down], latency=0(0), attempts=0]Could not find any remote Racks for fallback
--
--
TIME TO RUN: 1 seconds
```

```bash
		DynoJedisClient dynoClient = new DynoJedisClient.Builder()
					.withApplicationName("MY_APP")
		            .withDynomiteClusterName("MY_CLUSTER")
		            .withCPConfig( new ConnectionPoolConfigurationImpl("MY_APP")
		            					.setPort(8101)
		            					.withTokenSupplier(testTokenMapSupplier)
		            					.setMaxConnsPerHost(100) 
		            					
		            					//.setPoolShutdownDelay(0)
		            					.setConnectTimeout(4)
		            					//.setFailOnStartupIfNoHosts(true)
		            					//.setFailOnStartupIfNoHostsSeconds(12)
	                					//.setMaxTimeoutWhenExhausted(12)
		            					//.setSocketTimeout(12)
		            					.setRetryPolicyFactory(new RetryNTimes.RetryFactory(2))
		            )
		            .withHostSupplier(customHostSupplier)
		            .build();
```
```bash
log4j:WARN No appenders could be found for logger (com.netflix.dyno.jedis.DynoJedisClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
com.netflix.dyno.connectionpool.exception.NoAvailableHostsException: NoAvailableHostsException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: null, datacenter: null, status: Down], latency=0(0), attempts=0]Could not find any remote Racks for fallback
--
--
TIME TO RUN: 2 seconds
```


```bash
		DynoJedisClient dynoClient = new DynoJedisClient.Builder()
					.withApplicationName("MY_APP")
		            .withDynomiteClusterName("MY_CLUSTER")
		            .withCPConfig( new ConnectionPoolConfigurationImpl("MY_APP")
		            					.setPort(8101)
		            					.withTokenSupplier(testTokenMapSupplier)
		            					.setMaxConnsPerHost(100) 
		            					
		            					//.setPoolShutdownDelay(0)
		            					.setConnectTimeout(8)
		            					//.setFailOnStartupIfNoHosts(true)
		            					//.setFailOnStartupIfNoHostsSeconds(12)
	                					//.setMaxTimeoutWhenExhausted(12)
		            					//.setSocketTimeout(12)
		            					.setRetryPolicyFactory(new RetryNTimes.RetryFactory(2))
		            )
		            .withHostSupplier(customHostSupplier)
		            .build();
```
```bash
log4j:WARN No appenders could be found for logger (com.netflix.dyno.jedis.DynoJedisClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
com.netflix.dyno.connectionpool.exception.NoAvailableHostsException: NoAvailableHostsException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: null, datacenter: null, status: Down], latency=0(0), attempts=0]Could not find any remote Racks for fallback
--
--
TIME TO RUN: 3 seconds

```


```bash
		DynoJedisClient dynoClient = new DynoJedisClient.Builder()
					.withApplicationName("MY_APP")
		            .withDynomiteClusterName("MY_CLUSTER")
		            .withCPConfig( new ConnectionPoolConfigurationImpl("MY_APP")
		            					.setPort(8101)
		            					.withTokenSupplier(testTokenMapSupplier)
		            					.setMaxConnsPerHost(100) 
		            					
		            					//.setPoolShutdownDelay(0)
		            					.setConnectTimeout(16)
		            					//.setFailOnStartupIfNoHosts(true)
		            					//.setFailOnStartupIfNoHostsSeconds(12)
	                					//.setMaxTimeoutWhenExhausted(12)
		            					//.setSocketTimeout(12)
		            					.setRetryPolicyFactory(new RetryNTimes.RetryFactory(2))
		            )
		            .withHostSupplier(customHostSupplier)
		            .build();
```
```bash
log4j:WARN No appenders could be found for logger (com.netflix.dyno.jedis.DynoJedisClient).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
com.netflix.dyno.connectionpool.exception.NoAvailableHostsException: NoAvailableHostsException: [host=Host [hostname=UNKNOWN, ipAddress=UNKNOWN, port=0, rack: null, datacenter: null, status: Down], latency=0(0), attempts=0]Could not find any remote Racks for fallback
--
--
TIME TO RUN: 6 seconds

```

Cheers,
Diego Pacheco

**Comment by jcacciatore** [comment]
I will take a closer look

**Comment by diegopacheco** [comment]
@jcacciatore 

I was doing some investigations and I think I understand whats going on here. 

It's all about the HostConnectionPoolImpl class.

There is a default ThreadPool executor default 10. If you you want to prime more than 10 connections you might need to wait since only 10 will happen in parallel. For my case I was using 100 connections which I agree is too much 25 as you said could be better however we still have waiting and blocking here. 

Tell me if this makes sense to you:

The Math is CONNECTIONS_PER_HOST * RETRY_COUNT == TIME to RUN. We might have some flip here given network or even time to schedule and run the threads since default ThreadPool is 10. Also default Retry policy is 3. However you will try to do it first and them retry so in other words we will make by default 4 attempts to each connection per host.

I had this config and got this results which all makes sense and they are consistent to me:
```bash
setMaxConnsPerHost = 1
setConnectTimeout  = 1000
RUN TIME / TIMOUT  = 2s

setMaxConnsPerHost = 2
setConnectTimeout  = 1000
RUN TIME / TIMOUT  = 6s

setMaxConnsPerHost = 3
setConnectTimeout  = 1000
RUN TIME / TIMOUT  = 12s

setMaxConnsPerHost = 4
setConnectTimeout  = 1000
RUN TIME / TIMOUT  = 14s

setMaxConnsPerHost = 5
setConnectTimeout  = 1000
RUN TIME / TIMOUT  = 20s
```

My expectation was to have some kind of global timeout no matter how many connections per host I have BUT this config is not on dyno. I my sample with 100 connections I got 36s as timeout which is kind of gig number. 

So the timeout works but not in the way I was expecting because I was expecting a general timeout. So In the way I see this there is a couple of way we could have this behavior. 

1. Add this feature to dyno. Let's call main / general timeout.
2. Use hystrix and control this outside dyno. So If the timeout happens I kill the thread and all depending objects.

Does this make sense to you? What do you think would be the more reasonable solution?

**Comment by jcacciatore** [comment]
That is mostly correct. The retry policy for priming connections is fixed and currently not configurable by design. It's difficult to formalize a simple expression for time time b/c there are too many variables that are not in the client's control (e.g. thread scheduling). Generally speaking your math is a good rough estimate.  

As for a global timeout, this is not dyno client's responsibility and is better left to a library such as Hystrix. This is a common solution here at Netflix and works well.

Hope that helps

**Comment by diegopacheco** [comment]
@jcacciatore 

Agreed. 

Please close this issue. 

Thanks for you time.

Cheers,
Diego Pacheco

**Comment by jcacciatore** [comment]
@diegopacheco - Great suggestion; I completely agree that documenting this in the wiki would be valuable to others. I will update it. Thanks for your work on this !
